File: ios/PiStick/PiStickAPI.swift
import Foundation

@MainActor
final class PiStickAPI {
    private let config: ConfigStore
    private let state: StateStore
    private let tmdb: TMDBClient

    init(config: ConfigStore = ConfigStore(), state: StateStore = StateStore()) {
        self.config = config
        self.state = state
        self.tmdb = TMDBClient { config.credential }
    }

    func handle(path rawPath: String, method rawMethod: String, body: JSONObject) async throws -> JSONObject {
        guard let url = URL(string: "https://pistick.invalid\(rawPath)"),
              let components = URLComponents(url: url, resolvingAgainstBaseURL: false) else {
            throw PiStickFailure("The app request is invalid.")
        }
        let method = rawMethod.uppercased()
        let parts = components.path.split(separator: "/").map(String.init)
        let query = Dictionary(
            components.queryItems?.map { ($0.name, $0.value ?? "") } ?? [],
            uniquingKeysWith: { first, _ in first }
        )

        if parts == ["api", "status"], method == "GET" {
            return [
                "name": "PiStick",
                "version": Bundle.main.object(forInfoDictionaryKey: "CFBundleShortVersionString") as? String ?? "1.0.0",
                "tmdb_configured": config.isConfigured
            ]
        }

        if parts == ["api", "profiles"], method == "GET" {
            return state.profilesPayload()
        }
        if parts == ["api", "profiles"], method == "POST" {
            return ["profile": try state.addProfile(name: JSONTools.string(body["name"]))]
        }
        if parts.count == 3, parts[0] == "api", parts[1] == "profiles", method == "PATCH" {
            return ["profile": try state.renameProfile(parts[2], name: JSONTools.string(body["name"]))]
        }
        if parts.count == 3, parts[0] == "api", parts[1] == "profiles", method == "DELETE" {
            try state.deleteProfile(parts[2])
            return ["ok": true]
        }
        if parts.count == 4,
           parts[0] == "api", parts[1] == "profiles",
           parts[3] == "activate",
           method == "POST" {
            return ["profile": try state.activateProfile(parts[2])]
        }

        if parts == ["api", "settings", "tmdb"], method == "POST" {
            let credential = JSONTools.string(body["token"])
            guard ConfigStore.isUsable(credential) else {
                throw PiStickFailure("Paste your TMDB API Read Access Token or v3 API key.")
            }
            try await tmdb.validate(credential)
            try config.save(credential)
            return ["ok": true, "tmdb_configured": true]
        }

        if parts == ["api", "home"], method == "GET" {
            let profileID = try requiredProfileID(query["profile_id"])
            let payload = try await tmdb.home()
            return try decorateHome(profileID: profileID, payload: payload)
        }
        if parts.count == 3,
           parts[0] == "api", parts[1] == "discover",
           (parts[2] == "movie" || parts[2] == "tv"),
           method == "GET" {
            let profileID = try requiredProfileID(query["profile_id"])
            let page = try integer(query["page"] ?? "1", name: "Page", minimum: 1, maximum: 500)
            var payload = try await tmdb.discover(mediaType: parts[2], page: page)
            payload["items"] = try JSONTools.objects(payload["items"]).map {
                try state.decorate(profileID: profileID, media: $0)
            }
            return payload
        }
        if parts == ["api", "search"], method == "GET" {
            let profileID = try requiredProfileID(query["profile_id"])
            let page = try integer(query["page"] ?? "1", name: "Page", minimum: 1, maximum: 500)
            var payload = try await tmdb.search(query: query["q"] ?? "", page: page)
            payload["items"] = try JSONTools.objects(payload["items"]).map {
                try state.decorate(profileID: profileID, media: $0)
            }
            return payload
        }

        if parts.count == 4,
           parts[0] == "api", parts[1] == "media",
           (parts[2] == "movie" || parts[2] == "tv"),
           method == "GET" {
            let profileID = try requiredProfileID(query["profile_id"])
            let id = try integer(parts[3], name: "TMDB ID", minimum: 1)
            let media = try await tmdb.details(mediaType: parts[2], id: id)
            var decorated = try state.decorate(profileID: profileID, media: media)
            if parts[2] == "tv" {
                let resume = try state.resumeEpisode(profileID: profileID, media: media)
                decorated["resume_episode"] = ["season_number": resume.0, "episode_number": resume.1]
            }
            return ["media": decorated]
        }

        if parts.count == 5,
           parts[0] == "api", parts[1] == "tv", parts[3] == "season",
           method == "GET" {
            let profileID = try requiredProfileID(query["profile_id"])
            let showID = try integer(parts[2], name: "TMDB ID", minimum: 1)
            let seasonNumber = try integer(parts[4], name: "Season", minimum: 0, maximum: 10_000)
            return try await seasonPayload(showID: showID, seasonNumber: seasonNumber, profileID: profileID)
        }

        if parts == ["api", "continue"], method == "GET" {
            return ["items": try state.continueWatching(profileID: requiredProfileID(query["profile_id"]))]
        }

        if parts == ["api", "play"], method == "POST" {
            let profileID = try requiredProfileID(JSONTools.string(body["profile_id"]))
            let media = try requiredMedia(body["media"])
            let kind = JSONTools.string(media["media_type"])
            let mediaID = try integer(media["id"], name: "TMDB ID", minimum: 1)
            if kind == "movie" {
                let saved = try state.entry(profileID: profileID, media: media) ?? [:]
                let resume = JSONTools.double(saved["position_seconds"]) ?? 0
                try state.markStarted(profileID: profileID, media: media)
                return [
                    "embed_url": playbackURL(path: "/movie/\(mediaID)", resumeSeconds: resume),
                    "resume_seconds": resume,
                    "kind": "movie"
                ]
            }
            guard let episode = JSONTools.object(body["episode"]) else {
                throw PiStickFailure("Choose an episode first.")
            }
            let seasonNumber = try integer(episode["season_number"], name: "Season", minimum: 0, maximum: 10_000)
            let episodeNumber = try integer(episode["episode_number"], name: "Episode", minimum: 1, maximum: 100_000)
            let saved = try state.episodeEntry(
                profileID: profileID,
                media: media,
                seasonNumber: seasonNumber,
                episodeNumber: episodeNumber
            ) ?? [:]
            let resume = JSONTools.double(saved["position_seconds"]) ?? 0
            try state.markEpisodeStarted(profileID: profileID, media: media, episode: episode)
            return [
                "embed_url": playbackURL(path: "/tv/\(mediaID)/\(seasonNumber)/\(episodeNumber)", resumeSeconds: resume),
                "resume_seconds": resume,
                "kind": "episode",
                "season_number": seasonNumber,
                "episode_number": episodeNumber
            ]
        }

        if parts == ["api", "watch", "progress"], method == "POST" {
            let profileID = try requiredProfileID(JSONTools.string(body["profile_id"]))
            let media = try requiredMedia(body["media"])
            let position = try number(body["position_seconds"], name: "Position")
            let duration = try number(body["duration_seconds"], name: "Duration")
            if let episode = JSONTools.object(body["episode"]) {
                try state.setEpisodePosition(
                    profileID: profileID,
                    media: media,
                    episode: episode,
                    positionSeconds: position,
                    durationSeconds: duration
                )
            } else {
                try state.setPosition(
                    profileID: profileID,
                    media: media,
                    positionSeconds: position,
                    durationSeconds: duration
                )
            }
            return ["ok": true]
        }

        if parts == ["api", "watch", "action"], method == "POST" {
            let profileID = try requiredProfileID(JSONTools.string(body["profile_id"]))
            let media = try requiredMedia(body["media"])
            switch JSONTools.string(body["action"]) {
            case "finished", "show_finished":
                try state.markFinished(profileID: profileID, media: media)
            case "unwatched":
                try state.markUnwatched(profileID: profileID, media: media)
            case "started":
                try state.markStarted(profileID: profileID, media: media)
            case "episode_finished":
                guard let episode = JSONTools.object(body["episode"]) else {
                    throw PiStickFailure("Episode data is invalid.")
                }
                try state.markEpisodeFinished(profileID: profileID, media: media, episode: episode)
            default:
                throw PiStickFailure("Watch-state action is invalid.")
            }
            return ["ok": true]
        }

        throw PiStickFailure("App route not found.")
    }

    private func decorateHome(profileID: String, payload: JSONObject) throws -> JSONObject {
        var rows: [JSONObject] = []
        let continuing = try state.continueWatching(profileID: profileID)
        if !continuing.isEmpty { rows.append(["title": "Continue Watching", "items": continuing]) }
        for row in JSONTools.objects(payload["rows"]) {
            let items = try JSONTools.objects(row["items"]).map {
                try state.decorate(profileID: profileID, media: $0)
            }
            let title = JSONTools.string(row["title"])
            rows.append(["title": title.isEmpty ? "Explore" : title, "items": items])
        }
        let hero: Any
        if let item = payload["hero"] as? JSONObject {
            hero = try state.decorate(profileID: profileID, media: item)
        } else {
            hero = NSNull()
        }
        return ["hero": hero, "rows": rows]
    }

    private func seasonPayload(showID: Int, seasonNumber: Int, profileID: String) async throws -> JSONObject {
        let show = try await tmdb.details(mediaType: "tv", id: showID)
        var season = try await tmdb.season(showID: showID, seasonNumber: seasonNumber)
        season["episodes"] = try JSONTools.objects(season["episodes"]).map { episode in
            var result = episode
            let selectedSeason = JSONTools.int(episode["season_number"]) ?? seasonNumber
            let selectedEpisode = JSONTools.int(episode["episode_number"]) ?? 0
            if let watch = try state.episodeEntry(
                profileID: profileID,
                media: show,
                seasonNumber: selectedSeason,
                episodeNumber: selectedEpisode
            ) {
                result["watch"] = watch.filter {
                    ["status", "progress", "position_seconds", "duration_seconds"].contains($0.key)
                }
            }
            return result
        }
        return ["season": season]
    }

    private func requiredProfileID(_ candidate: String?) throws -> String {
        let requested = (candidate ?? "").trimmingCharacters(in: .whitespacesAndNewlines)
        let selected = requested.isEmpty ? (state.activeProfileID ?? "") : requested
        guard state.profile(selected) != nil else { throw PiStickFailure("Choose a profile first.") }
        return selected
    }

    private func requiredMedia(_ candidate: Any?) throws -> JSONObject {
        guard let media = JSONTools.object(candidate) else { throw PiStickFailure("A media object is required.") }
        let kind = JSONTools.string(media["media_type"])
        guard kind == "movie" || kind == "tv", (JSONTools.int(media["id"]) ?? 0) > 0 else {
            throw PiStickFailure("Media must contain a valid TMDB ID and type.")
        }
        return media
    }

    private func integer(
        _ candidate: Any?,
        name: String,
        minimum: Int,
        maximum: Int = 2_147_483_647
    ) throws -> Int {
        guard let value = JSONTools.int(candidate), value >= minimum, value <= maximum else {
            throw PiStickFailure("\(name) is outside the allowed range.")
        }
        return value
    }

    private func number(_ candidate: Any?, name: String) throws -> Double {
        guard let value = JSONTools.double(candidate), value.isFinite, value >= 0, value <= 10_000_000 else {
            throw PiStickFailure("\(name) must be a non-negative number.")
        }
        return value
    }

    private func playbackURL(path: String, resumeSeconds: Double) -> String {
        let base = "https://player.videasy.to\(path)"
        let seconds = max(0, Int(resumeSeconds))
        return seconds > 0 ? "\(base)?progress=\(seconds)" : base
    }
}

File: android/app/src/main/java/app/pistick/android/PiStickApi.java
package app.pistick.android;

import android.content.Context;

import org.json.JSONArray;
import org.json.JSONObject;

import java.io.UnsupportedEncodingException;
import java.net.URI;
import java.net.URISyntaxException;
import java.net.URLDecoder;
import java.util.ArrayList;
import java.util.LinkedHashMap;
import java.util.List;
import java.util.Locale;
import java.util.Map;

final class PiStickApi implements AutoCloseable {
    private final CredentialStore credentials;
    private final StateStore state;
    private final TmdbClient tmdb;

    PiStickApi(Context context) {
        credentials = new CredentialStore(context);
        state = new StateStore(context);
        tmdb = new TmdbClient(credentials::credential);
    }

    JSONObject handle(String rawPath, String rawMethod, JSONObject body) throws PiStickException {
        RequestTarget target = RequestTarget.parse(rawPath);
        List<String> parts = target.parts;
        Map<String, String> query = target.query;
        String method = rawMethod == null || rawMethod.trim().isEmpty()
                ? "GET"
                : rawMethod.toUpperCase(Locale.ROOT);
        JSONObject requestBody = body == null ? new JSONObject() : body;

        if (matches(parts, "api", "status") && "GET".equals(method)) {
            JSONObject result = new JSONObject();
            JsonUtils.put(result, "name", "PiStick");
            JsonUtils.put(result, "version", BuildConfig.VERSION_NAME);
            JsonUtils.put(result, "tmdb_configured", credentials.isConfigured());
            return result;
        }

        if (matches(parts, "api", "profiles") && "GET".equals(method)) {
            return state.profilesPayload();
        }
        if (matches(parts, "api", "profiles") && "POST".equals(method)) {
            return JsonUtils.put(new JSONObject(), "profile", state.addProfile(requestBody.optString("name", "")));
        }
        if (parts.size() == 3 && "api".equals(parts.get(0)) && "profiles".equals(parts.get(1))
                && "PATCH".equals(method)) {
            return JsonUtils.put(new JSONObject(), "profile",
                    state.renameProfile(parts.get(2), requestBody.optString("name", "")));
        }
        if (parts.size() == 3 && "api".equals(parts.get(0)) && "profiles".equals(parts.get(1))
                && "DELETE".equals(method)) {
            state.deleteProfile(parts.get(2));
            return JsonUtils.put(new JSONObject(), "ok", true);
        }
        if (parts.size() == 4 && "api".equals(parts.get(0)) && "profiles".equals(parts.get(1))
                && "activate".equals(parts.get(3)) && "POST".equals(method)) {
            return JsonUtils.put(new JSONObject(), "profile", state.activateProfile(parts.get(2)));
        }

        if (matches(parts, "api", "settings", "tmdb") && "POST".equals(method)) {
            String credential = requestBody.optString("token", "");
            if (!CredentialStore.isUsable(credential)) {
                throw new PiStickException("Paste your TMDB API Read Access Token or v3 API key.");
            }
            tmdb.validate(credential);
            credentials.save(credential);
            tmdb.clearCache();
            JSONObject result = new JSONObject();
            JsonUtils.put(result, "ok", true);
            JsonUtils.put(result, "tmdb_configured", true);
            return result;
        }

        if (matches(parts, "api", "home") && "GET".equals(method)) {
            String profileId = requiredProfileId(query.get("profile_id"));
            return decorateHome(profileId, tmdb.home());
        }
        if (parts.size() == 3 && "api".equals(parts.get(0)) && "discover".equals(parts.get(1))
                && ("movie".equals(parts.get(2)) || "tv".equals(parts.get(2))) && "GET".equals(method)) {
            String profileId = requiredProfileId(query.get("profile_id"));
            int page = integer(query.getOrDefault("page", "1"), "Page", 1, 500);
            JSONObject payload = tmdb.discover(parts.get(2), page);
            JsonUtils.put(payload, "items", decorateItems(profileId, payload.optJSONArray("items")));
            return payload;
        }
        if (matches(parts, "api", "search") && "GET".equals(method)) {
            String profileId = requiredProfileId(query.get("profile_id"));
            int page = integer(query.getOrDefault("page", "1"), "Page", 1, 500);
            JSONObject payload = tmdb.search(query.getOrDefault("q", ""), page);
            JsonUtils.put(payload, "items", decorateItems(profileId, payload.optJSONArray("items")));
            return payload;
        }

        if (parts.size() == 4 && "api".equals(parts.get(0)) && "media".equals(parts.get(1))
                && ("movie".equals(parts.get(2)) || "tv".equals(parts.get(2))) && "GET".equals(method)) {
            String profileId = requiredProfileId(query.get("profile_id"));
            int id = integer(parts.get(3), "TMDB ID", 1, Integer.MAX_VALUE);
            JSONObject media = tmdb.details(parts.get(2), id);
            JSONObject decorated = state.decorate(profileId, media);
            if ("tv".equals(parts.get(2))) {
                int[] resume = state.resumeEpisode(profileId, media);
                JSONObject resumeJson = new JSONObject();
                JsonUtils.put(resumeJson, "season_number", resume[0]);
                JsonUtils.put(resumeJson, "episode_number", resume[1]);
                JsonUtils.put(decorated, "resume_episode", resumeJson);
            }
            return JsonUtils.put(new JSONObject(), "media", decorated);
        }

        if (parts.size() == 5 && "api".equals(parts.get(0)) && "tv".equals(parts.get(1))
                && "season".equals(parts.get(3)) && "GET".equals(method)) {
            String profileId = requiredProfileId(query.get("profile_id"));
            int showId = integer(parts.get(2), "TMDB ID", 1, Integer.MAX_VALUE);
            int seasonNumber = integer(parts.get(4), "Season", 0, 10_000);
            return seasonPayload(showId, seasonNumber, profileId);
        }

        if (matches(parts, "api", "continue") && "GET".equals(method)) {
            return JsonUtils.put(new JSONObject(), "items",
                    state.continueWatching(requiredProfileId(query.get("profile_id"))));
        }

        if (matches(parts, "api", "play") && "POST".equals(method)) {
            String profileId = requiredProfileId(requestBody.optString("profile_id", ""));
            JSONObject media = requiredMedia(requestBody.opt("media"));
            String kind = media.optString("media_type", "");
            int mediaId = integer(media.opt("id"), "TMDB ID", 1, Integer.MAX_VALUE);
            if ("movie".equals(kind)) {
                JSONObject saved = state.entry(profileId, media);
                double resume = JsonUtils.number(saved.opt("position_seconds"), 0);
                state.markStarted(profileId, media);
                JSONObject result = new JSONObject();
                JsonUtils.put(result, "embed_url", playbackUrl("/movie/" + mediaId, resume));
                JsonUtils.put(result, "resume_seconds", resume);
                JsonUtils.put(result, "kind", "movie");
                return result;
            }
            JSONObject episode = requestBody.optJSONObject("episode");
            if (episode == null) throw new PiStickException("Choose an episode first.");
            int seasonNumber = integer(episode.opt("season_number"), "Season", 0, 10_000);
            int episodeNumber = integer(episode.opt("episode_number"), "Episode", 1, 100_000);
            JSONObject saved = state.episodeEntry(profileId, media, seasonNumber, episodeNumber);
            double resume = JsonUtils.number(saved.opt("position_seconds"), 0);
            state.markEpisodeStarted(profileId, media, episode);
            JSONObject result = new JSONObject();
            JsonUtils.put(result, "embed_url",
                    playbackUrl("/tv/" + mediaId + "/" + seasonNumber + "/" + episodeNumber, resume));
            JsonUtils.put(result, "resume_seconds", resume);
            JsonUtils.put(result, "kind", "episode");
            JsonUtils.put(result, "season_number", seasonNumber);
            JsonUtils.put(result, "episode_number", episodeNumber);
            return result;
        }

        if (matches(parts, "api", "watch", "progress") && "POST".equals(method)) {
            String profileId = requiredProfileId(requestBody.optString("profile_id", ""));
            JSONObject media = requiredMedia(requestBody.opt("media"));
            double position = number(requestBody.opt("position_seconds"), "Position");
            double duration = number(requestBody.opt("duration_seconds"), "Duration");
            JSONObject episode = requestBody.optJSONObject("episode");
            if (episode == null) state.setPosition(profileId, media, position, duration);
            else state.setEpisodePosition(profileId, media, episode, position, duration);
            return JsonUtils.put(new JSONObject(), "ok", true);
        }

        if (matches(parts, "api", "watch", "action") && "POST".equals(method)) {
            String profileId = requiredProfileId(requestBody.optString("profile_id", ""));
            JSONObject media = requiredMedia(requestBody.opt("media"));
            switch (requestBody.optString("action", "")) {
                case "finished":
                case "show_finished":
                    state.markFinished(profileId, media);
                    break;
                case "unwatched":
                    state.markUnwatched(profileId, media);
                    break;
                case "started":
                    state.markStarted(profileId, media);
                    break;
                case "episode_finished":
                    JSONObject episode = requestBody.optJSONObject("episode");
                    if (episode == null) throw new PiStickException("Episode data is invalid.");
                    state.markEpisodeFinished(profileId, media, episode);
                    break;
                default:
                    throw new PiStickException("Watch-state action is invalid.");
            }
            return JsonUtils.put(new JSONObject(), "ok", true);
        }

        throw new PiStickException("App route not found.");
    }

    @Override
    public void close() {
        tmdb.close();
    }

    private JSONObject decorateHome(String profileId, JSONObject payload) throws PiStickException {
        JSONArray rows = new JSONArray();
        JSONArray continuing = state.continueWatching(profileId);
        if (continuing.length() > 0) {
            JSONObject row = new JSONObject();
            JsonUtils.put(row, "title", "Continue Watching");
            JsonUtils.put(row, "items", continuing);
            rows.put(row);
        }
        JSONArray sourceRows = payload.optJSONArray("rows");
        if (sourceRows != null) {
            for (int index = 0; index < sourceRows.length(); index++) {
                JSONObject source = sourceRows.optJSONObject(index);
                if (source == null) continue;
                JSONObject row = new JSONObject();
                String title = source.optString("title", "");
                JsonUtils.put(row, "title", title.isEmpty() ? "Explore" : title);
                JsonUtils.put(row, "items", decorateItems(profileId, source.optJSONArray("items")));
                rows.put(row);
            }
        }
        Object rawHero = payload.opt("hero");
        Object hero = JSONObject.NULL;
        if (rawHero instanceof JSONObject) hero = state.decorate(profileId, (JSONObject) rawHero);
        JSONObject result = new JSONObject();
        JsonUtils.put(result, "hero", hero);
        JsonUtils.put(result, "rows", rows);
        return result;
    }

    private JSONObject seasonPayload(int showId, int seasonNumber, String profileId) throws PiStickException {
        JSONObject show = tmdb.details("tv", showId);
        JSONObject season = tmdb.season(showId, seasonNumber);
        JSONArray decorated = new JSONArray();
        JSONArray episodes = season.optJSONArray("episodes");
        if (episodes != null) {
            for (int index = 0; index < episodes.length(); index++) {
                JSONObject episode = JsonUtils.copy(episodes.optJSONObject(index));
                int selectedSeason = episode.optInt("season_number", seasonNumber);
                int selectedEpisode = episode.optInt("episode_number", 0);
                JSONObject watch = state.episodeEntry(profileId, show, selectedSeason, selectedEpisode);
                if (watch.length() > 0) {
                    JsonUtils.put(episode, "watch", JsonUtils.selected(watch,
                            "status", "progress", "position_seconds", "duration_seconds"));
                }
                decorated.put(episode);
            }
        }
        JsonUtils.put(season, "episodes", decorated);
        return JsonUtils.put(new JSONObject(), "season", season);
    }

    private JSONArray decorateItems(String profileId, JSONArray items) throws PiStickException {
        JSONArray result = new JSONArray();
        if (items == null) return result;
        for (int index = 0; index < items.length(); index++) {
            JSONObject item = items.optJSONObject(index);
            if (item != null) result.put(state.decorate(profileId, item));
        }
        return result;
    }

    private String requiredProfileId(String candidate) throws PiStickException {
        String requested = candidate == null ? "" : candidate.trim();
        String selected = requested.isEmpty() ? state.activeProfileId() : requested;
        if (selected == null || state.profile(selected) == null) {
            throw new PiStickException("Choose a profile first.");
        }
        return selected;
    }

    private JSONObject requiredMedia(Object candidate) throws PiStickException {
        JSONObject media = candidate instanceof JSONObject ? (JSONObject) candidate : null;
        if (media == null) throw new PiStickException("A media object is required.");
        String kind = media.optString("media_type", "");
        if (!("movie".equals(kind) || "tv".equals(kind)) || media.optInt("id", 0) <= 0) {
            throw new PiStickException("Media must contain a valid TMDB ID and type.");
        }
        return media;
    }

    private int integer(Object candidate, String name, int minimum, int maximum) throws PiStickException {
        double parsed = JsonUtils.number(candidate, Double.NaN);
        if (!Double.isFinite(parsed) || Math.rint(parsed) != parsed || parsed < minimum || parsed > maximum) {
            throw new PiStickException(name + " is outside the allowed range.");
        }
        return (int) parsed;
    }

    private double number(Object candidate, String name) throws PiStickException {
        double value = JsonUtils.number(candidate, Double.NaN);
        if (!Double.isFinite(value) || value < 0 || value > 10_000_000) {
            throw new PiStickException(name + " must be a non-negative number.");
        }
        return value;
    }

    static String playbackUrl(String path, double resumeSeconds) {
        int seconds = Math.max(0, (int) resumeSeconds);
        StringBuilder url = new StringBuilder("https://player.videasy.to").append(path);
        url.append("?autoplay=true");
        if (path.startsWith("/tv/")) url.append("&autoplayNextEpisode=true");
        if (seconds > 0) url.append("&progress=").append(seconds);
        return url.toString();
    }

    private static boolean matches(List<String> parts, String... expected) {
        if (parts.size() != expected.length) return false;
        for (int index = 0; index < expected.length; index++) {
            if (!expected[index].equals(parts.get(index))) return false;
        }
        return true;
    }

    private static final class RequestTarget {
        final List<String> parts;
        final Map<String, String> query;

        RequestTarget(List<String> parts, Map<String, String> query) {
            this.parts = parts;
            this.query = query;
        }

        static RequestTarget parse(String rawPath) throws PiStickException {
            if (rawPath == null || rawPath.length() > 4096 || !rawPath.startsWith("/api/")) {
                throw new PiStickException("The app request is invalid.");
            }
            try {
                URI uri = new URI(rawPath);
                List<String> parts = new ArrayList<>();
                for (String piece : uri.getRawPath().split("/")) {
                    if (!piece.isEmpty()) parts.add(decode(piece));
                }
                Map<String, String> query = new LinkedHashMap<>();
                String rawQuery = uri.getRawQuery();
                if (rawQuery != null && !rawQuery.isEmpty()) {
                    for (String pair : rawQuery.split("&")) {
                        String[] values = pair.split("=", 2);
                        String key = decode(values[0]);
                        if (!query.containsKey(key)) {
                            query.put(key, values.length == 2 ? decode(values[1]) : "");
                        }
                    }
                }
                return new RequestTarget(parts, query);
            } catch (URISyntaxException | IllegalArgumentException error) {
                throw new PiStickException("The app request is invalid.", error);
            }
        }

        private static String decode(String value) {
            try {
                return URLDecoder.decode(value, "UTF-8");
            } catch (UnsupportedEncodingException error) {
                throw new IllegalStateException(error);
            }
        }
    }
}
